File: 108. Convert Sorted Array to Binary Search Tree/sortedArrayToBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function(nums) {
    if (!nums == null || !nums.length) {
        return null
    }
    else {
        let p = Math.floor(nums.length/2) //must define
        const n = new TreeNode(nums[p]) //must define
        n.left = sortedArrayToBST(nums.slice(0,p))
        n.right = sortedArrayToBST(nums.slice(p+1,nums.length))
        return n
    }
};

// Runtime 84 ms / Memory 44.4 MB
// debug: https://jsfiddle.net/

File: 108. Convert Sorted Array to Binary Search Tree/sortedArrayToBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        // empty
        if(nums.size()==0) { return NULL; }
        // root
        if(nums.size()==1) { return new TreeNode(nums[0]); }
        // binary search reverse
        int mid = nums.size()/2;
        TreeNode *root=new TreeNode(nums[mid]);
        vector<int> numsL(nums.begin(),nums.begin()+mid);
        vector<int> numsR(nums.begin()+mid+1,nums.end());
        root->left = sortedArrayToBST(numsL);
        root->right = sortedArrayToBST(numsR);
        return root;
    }
};